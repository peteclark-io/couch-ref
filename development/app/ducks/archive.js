'use strict';

import firebase from 'firebase';

import references from '../core/references';
import {createMatch, createQuestion, createStatistic} from '../core/mappers';

const LOAD_MATCH = 'couch-ref/archive/LOAD_MATCH'
const LOAD_QUESTION = 'couch-ref/archive/LOAD_QUESTION'
const LOAD_STATISTICS = 'couch-ref/archive/LOAD_STATISTICS'

export default function reducer(state = {}, action){
   switch(action.type){
      case LOAD_MATCH:
      return Object.assign({}, state, {
         matches: Object.assign({}, state.matches, {
            [action.match.id]: action.match
         })
      });

      case LOAD_QUESTION:
      return Object.assign({}, state, {
         questions: Object.assign({}, state.questions, {
            [action.question.id]: action.question
         })
      });

      case LOAD_STATISTICS:
      return Object.assign({}, state, {
         statistics: Object.assign({}, state.statistics, {
            [action.statistic.id]: action.statistic
         })
      });

      default:
      return state;
   }
};

export function loadMatch(match, redirect) {
   return (dispatch) => {
      var db = firebase.database();
      db.ref(references.archiveMatches + '/' + match).once('value').then((snap) => {
         if (!snap.exists()){
            redirect ? redirect() : null;
            return;
         }

         var archived = snap.val();

         dispatch({type: LOAD_MATCH, match: createMatch(archived)});

         if (!archived.questions){
            return;
         }

         archived.questions.map(q => {
            db.ref(references.archiveQuestions + '/' + q.id).once('value').then(snap => {
               var archivedQ = snap.val();
               dispatch({type: LOAD_QUESTION, question: createQuestion(archivedQ)});
            });
         });
      });
   };
}

export function loadQuestion(question, redirect) {
   return (dispatch) => {
      var db = firebase.database();
      db.ref(references.archiveQuestions + '/' + question).once('value').then((snap) => {
         if (!snap.exists()){
            redirect ? redirect() : null;
            return;
         }

         var archived = snap.val();
         dispatch({type: LOAD_QUESTION, question: createQuestion(archived)});
      });
   };
}

export function loadQuestionStatistics(question, redirect) {
   return (dispatch) => {
      var db = firebase.database();
      db.ref(references.archiveStatistics + '/' + question).once('value').then((snap) => {
         if (!snap.exists()){
            redirect ? redirect() : null;
            return;
         }

         var archived = snap.val();
         dispatch({type: LOAD_STATISTICS, statistic: createStatistic(archived)});

         db.ref(references.archiveQuestions + '/' + archived.id).once('value').then(snap => {
            var archivedQ = snap.val();
            loadMatch(archivedQ.match)(dispatch);
         });
      });
   };
}
