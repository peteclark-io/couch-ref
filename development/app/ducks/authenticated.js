'use strict';

const USER_AUTHENTICATED = 'couch-ref/authenticated/USER_AUTHENTICATED';

export default function reducer(state = false, action){
  switch(action.type){
    case USER_AUTHENTICATED:
      return action.authenticated; // should this just be true?
    default:
      return state;
  }
};

export function authenticated() {
  return {type: USER_AUTHENTICATED, authenticated: true};
};
