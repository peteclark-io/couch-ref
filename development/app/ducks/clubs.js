'use strict';

import _ from 'lodash';

const ADD_CLUBS = 'couch-ref/clubs/ADD_CLUBS'

export default function reducer(state = [], action){
   switch(action.type){
      case ADD_CLUBS:
      return _.sortBy(action.clubs, ['shortName']);

      default:
      return state;
   }
};

export function addClubs(clubs) {
   return {type: ADD_CLUBS, clubs: clubs};
}
