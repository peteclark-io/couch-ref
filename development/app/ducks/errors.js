'use strict';

const ADD_ERROR = 'couch-ref/errors/ADD_ERROR'

export default function reducer(state = {}, action){
  switch(action.type){
    case ADD_ERROR:
      return Object.assign({}, state, {
         [action.code]: {
            code: action.code,
            message: action.msg
         }
      });

    default:
      return state;
  }
};

export function addError(msg, code) {
  return {type: ADD_ERROR, msg: msg, code: code};
}
