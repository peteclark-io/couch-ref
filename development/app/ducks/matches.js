'use strict'

import _ from 'lodash';

const MATCH_UPDATE = 'couch-ref/matches/MATCH_UPDATE'
const ADD_MATCH = 'couch-ref/matches/ADD_MATCH'

function sort(arr){
  return _.sortBy(arr, ['fullTime', 'kickOff']);
}

export default function reducer(state = [], action){
  var match = action.match;

  switch(action.type){
    case MATCH_UPDATE:
      console.log('Action found!', match);
      return sort([
        ..._.differenceWith(state, [match], (val, compare) => {return val.id === compare.id}),
        match
      ]);

    case ADD_MATCH:
      console.log('Action found!', match);
      var update = state.filter((entry) => { return entry.id === match.id; }).length > 0;
      console.log('Is update', update);
      if (update){
         return sort([
           ..._.differenceWith(state, [match], (val, compare) => {return val.id === compare.id}),
           match
         ]);
      }

      return sort([...state, match]);
    default:
      return state;
  }
};

export function updateMatch(match) {
  return {type: MATCH_UPDATE, match: match};
}

export function addMatch(match) {
  return {type: ADD_MATCH, match: match};
}
