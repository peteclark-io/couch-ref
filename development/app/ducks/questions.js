'use strict';

const QUESTION_UPDATE = 'couch-ref/questions/MATCH_UPDATE'
const ADD_QUESTION = 'couch-ref/questions/ADD_MATCH'

export default function reducer(state = {}, action){
  switch(action.type){
    case QUESTION_UPDATE:
      return Object.assign({}, state, {
        [action.question.id]: action.question
      });

    case ADD_QUESTION:
      return Object.assign({}, state, {
        [action.question.id]: action.question
      });

    default:
      return state;
  }
};

export function updateQuestion(question) {
  return {type: QUESTION_UPDATE, question: question};
}

export function addQuestion(question) {
  return {type: ADD_QUESTION, question: question};
}
