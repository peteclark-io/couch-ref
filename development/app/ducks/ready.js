'use strict';

const APP_READY = 'couch-ref/ready/APP_READY';

export default function reducer(state = false, action){
  switch(action.type){
    case APP_READY:
      return action.ready; // should this just be true?
    default:
      return state;
  }
};

export function ready() {
  return {type: APP_READY, ready: true};
};
