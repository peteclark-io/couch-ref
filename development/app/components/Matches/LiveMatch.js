'use strict';

import { connect } from 'react-redux';
import _ from 'lodash';

import {vote} from '../../ducks/user';
import {saveVote} from '../../core/db-actions';
import {saveVoteAsCookie} from '../../core/cookies';

import Match from './Match';

const getLiveMatch = (state = {matches: {}}, id) => {
   return !state.matches || !state.matches[id] ? undefined : state.matches[id];
};

const getScoresReady = (state = {user: {votes: {}}}, match) => {
   if (!match) {
      return false;
   }

   var scores = match.questions.map((q) => {
      if (!state.user.votes || !state.user.votes[q.id]){
         return undefined;
      }

      var answer = state.user.votes[q.id];
      return answer ? answer.score : undefined;
   }).filter((s) => {return s;});

   return scores.length !== 0;
};

const getQuestions = (state = {questions: {}}, match) => {
   if (!state.questions || !match){
      return undefined;
   }

   var reduced = match.questions.map(q => {
      return state.questions[q.id] ? state.questions[q.id] : undefined;
   }).filter(q => q);

   if (reduced.length === 0){
      return undefined;
   }

   return _.keyBy(reduced, 'id');
};

const getUser = (state = {user: {}}) => {
   return state.user;
};

const getStatistics = (state = {statistics: {}}, match) => {
   if (!state.statistics || !match){
      return undefined;
   }

   var reduced = match.questions.map(q => {
      return state.statistics[q.id] ? state.statistics[q.id] : undefined;
   }).filter(q => q);

   if (reduced.length === 0){
      return undefined;
   }

   return _.keyBy(reduced, 'id');
};

const getReferee = (state = {referees: {}}, match) => {
   if (!match){
      return undefined;
   }

   var ref = state.referees[match.referee];
   return ref ? ref : undefined;
};

const mapStateToProps = (state, ownProps) => {
   var match = getLiveMatch(state, ownProps.id);
   return {
      match: match,
      questions: getQuestions(state, match),
      statistics: getStatistics(state, match),
      referee: getReferee(state, match),
      scoresReady: getScoresReady(state, match),
      user: getUser(state)
   };
};

const mapDispatchToProps = (dispatch) => {
   return {
      vote: (result, question, user) => {
         saveVote(user, question, result);
         saveVoteAsCookie(user, question, result);
         dispatch(vote(question, result));
      }
   }
};

const LiveMatch = connect(
   mapStateToProps,
   mapDispatchToProps
)(Match);

export default LiveMatch;
