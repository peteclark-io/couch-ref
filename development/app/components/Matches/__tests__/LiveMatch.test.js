'use strict';

import React from 'react';
import LiveMatch from '../LiveMatch';
import {shallow, mount} from 'enzyme';

import reducer from '../../../ducks/reducers';
import {addMatch, updateMatch} from '../../../ducks/matches';
import {createStore} from 'redux';
import {Provider} from 'react-redux';

it('Displays loader when there is no data.', () => {
   var store = createStore(reducer);
   const rendered = mount(
      <Provider store={store}>
         <LiveMatch id={'1001'} />
      </Provider>
   );

   expect(rendered.find('.loading')).toHaveLength(1);
});

it('Displays loader when there is no good data.', () => {
   var store = createStore(reducer);
   const rendered = mount(
      <Provider store={store}>
         <LiveMatch id={'1001'} />
      </Provider>
   );

   var match = {id: '1002', home: 'Arsenal', away: 'West Ham', goalsHome: 1, goalsAway: 0, questions: []};
   store.dispatch(addMatch(match));

   expect(rendered.find('.loading')).toHaveLength(1);
});

it('Displays something else when there is data, but no questions.', () => {
   var store = createStore(reducer);
   const rendered = mount(
      <Provider store={store}>
         <LiveMatch id={'1001'} />
      </Provider>
   );

   var match = {id: '1001', home: 'Arsenal', away: 'West Ham', goalsHome: 1, goalsAway: 0, questions: []};
   store.dispatch(addMatch(match));

   expect(rendered.html()).toMatchSnapshot();
});

it('Changes the score.', () => {
   var store = createStore(reducer);
   const rendered = mount(
      <Provider store={store}>
         <LiveMatch id={'1001'} />
      </Provider>
   );

   var match = {id: '1001', home: 'Arsenal', away: 'West Ham', goalsHome: 1, goalsAway: 0, questions: []};
   store.dispatch(addMatch(match));

   expect(rendered.html()).toMatchSnapshot();

   match = Object.assign({}, match, {goalsHome: 2})
   store.dispatch(updateMatch(match));

   expect(rendered.html()).toMatchSnapshot();
});
