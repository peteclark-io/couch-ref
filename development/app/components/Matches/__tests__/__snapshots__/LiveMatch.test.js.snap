exports[`test Changes the score. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12 col-sm-8\"><h2 class=\"match-header\"><!-- react-text: 10 -->Arsenal<!-- /react-text --><!-- react-text: 11 --> <!-- /react-text --><!-- react-text: 12 -->1<!-- /react-text --><!-- react-text: 13 -->  -  <!-- /react-text --><!-- react-text: 14 -->0<!-- /react-text --><!-- react-text: 15 --> <!-- /react-text --><!-- react-text: 16 -->West Ham<!-- /react-text --></h2></div></div><!-- react-empty: 17 --><!-- react-empty: 18 --><div class=\"questions-list\"><h1 class=\"header\">Live Questions</h1><div class=\"row\"><div class=\"col-xs-12\"><h4 class=\"spacer\">No questions have been asked yet</h4></div></div></div></div>"`;

exports[`test Changes the score. 2`] = `"<div><div class=\"row\"><div class=\"col-xs-12 col-sm-8\"><h2 class=\"match-header\"><!-- react-text: 10 -->Arsenal<!-- /react-text --><!-- react-text: 11 --> <!-- /react-text --><!-- react-text: 12 -->2<!-- /react-text --><!-- react-text: 13 -->  -  <!-- /react-text --><!-- react-text: 14 -->0<!-- /react-text --><!-- react-text: 15 --> <!-- /react-text --><!-- react-text: 16 -->West Ham<!-- /react-text --></h2></div></div><!-- react-empty: 17 --><!-- react-empty: 18 --><div class=\"questions-list\"><h1 class=\"header\">Live Questions</h1><div class=\"row\"><div class=\"col-xs-12\"><h4 class=\"spacer\">No questions have been asked yet</h4></div></div></div></div>"`;

exports[`test Displays something else when there is data, but no questions. 1`] = `"<div><div class=\"row\"><div class=\"col-xs-12 col-sm-8\"><h2 class=\"match-header\"><!-- react-text: 10 -->Arsenal<!-- /react-text --><!-- react-text: 11 --> <!-- /react-text --><!-- react-text: 12 -->1<!-- /react-text --><!-- react-text: 13 -->  -  <!-- /react-text --><!-- react-text: 14 -->0<!-- /react-text --><!-- react-text: 15 --> <!-- /react-text --><!-- react-text: 16 -->West Ham<!-- /react-text --></h2></div></div><!-- react-empty: 17 --><!-- react-empty: 18 --><div class=\"questions-list\"><h1 class=\"header\">Live Questions</h1><div class=\"row\"><div class=\"col-xs-12\"><h4 class=\"spacer\">No questions have been asked yet</h4></div></div></div></div>"`;
